src/NTRIP/NTRIP.cc

```diff
@@ -0,0 +1,172 @@
+ *
+ * QGroundControl is licensed according to the terms in the file
+ * COPYING.md in the root of the source code directory.
+ *
+ ****************************************************************************/
+
+#include "NTRIP.h"
+#include "QGCLoggingCategory.h"
+#include "QGCApplication.h"
+#include "SettingsManager.h"
+#include "NTRIPSettings.h"
+
+#include <QDebug>
+
+NTRIP::NTRIP(QGCApplication* app, QGCToolbox* toolbox)
+    : QGCTool(app, toolbox)
+{
+}
+
+void NTRIP::setToolbox(QGCToolbox* toolbox)
+{
+    QGCTool::setToolbox(toolbox);
+    
+    NTRIPSettings* settings = qgcApp()->toolbox()->settingsManager()->ntripSettings();
+    if (settings->ntripServerConnectEnabled()->rawValue().toBool()) {
+        _rtcmMavlink = new RTCMMavlink(*toolbox);
+        
+        _tcpLink = new NTRIPTCPLink(settings->ntripServerHostAddress()->rawValue().toString(),
+                                    settings->ntripServerPort()->rawValue().toInt(),
+                                    settings->ntripUsername()->rawValue().toString(),
+                                    settings->ntripPassword()->rawValue().toString(),
+                                    settings->ntripMountpoint()->rawValue().toString(),
+                                    settings->ntripWhitelist()->rawValue().toString(),
+                                    this);
+        connect(_tcpLink, &NTRIPTCPLink::error,              this, &NTRIP::_tcpError,           Qt::QueuedConnection);
+        connect(_tcpLink, &NTRIPTCPLink::RTCMDataUpdate,   _rtcmMavlink, &RTCMMavlink::RTCMDataUpdate);
+    }
+}
+
+
+void NTRIP::_tcpError(const QString errorMsg)
+{
+    qgcApp()->showAppMessage(tr("NTRIP Server Error: %1").arg(errorMsg));
+}
+
+
+NTRIPTCPLink::NTRIPTCPLink(const QString& hostAddress,
+                           int port,
+                           const QString &username,
+                           const QString &password,
+                           const QString &mountpoint,
+                           const QString &whitelist,
+                           QObject* parent)
+    : QThread       (parent)
+    , _hostAddress  (hostAddress)
+    , _port         (port)
+    , _username     (username)
+    , _password     (password)
+    , _mountpoint   (mountpoint)
+{
+    moveToThread(this);
+    start();
+    for(const auto& msg: whitelist.split(',')){
+        int msg_int = msg.toInt();
+        if(msg_int)
+            _whitelist.append(msg_int);
+    }
+    qCDebug(NTRIPLog) << "whitelist: " << _whitelist;
+    if (!_rtcm_parsing) {
+        _rtcm_parsing = new RTCMParsing();
+    }
+    _rtcm_parsing->reset();
+    _state = NTRIPState::uninitialised;
+}
+
+NTRIPTCPLink::~NTRIPTCPLink(void)
+{
+    if (_socket) {
+        QObject::disconnect(_socket, &QTcpSocket::readyRead, this, &NTRIPTCPLink::_readBytes);
+        _socket->disconnectFromHost();
+        _socket->deleteLater();
+        _socket = nullptr;
+    }
+    quit();
+    wait();
```

Comment:
Same here: quit & wait first, then cleanup.
Also there's a memory leak for _rtcm_parsing